internal/provider/resource_network.go

```diff
@@ -194,39 +243,73 @@ func NewNetworkDeployer(ctx context.Context, d *schema.ResourceData, apiClient *
-		APIClient:        apiClient,
-		ncPool:           pool,
-		deployer:         deployer.NewDeployer(apiClient.identity, apiClient.twin_id, apiClient.grid_client, pool, true, nil, string(deploymentDataStr)),
+		Name:                  d.Get("name").(string),
+		Description:           d.Get("description").(string),
+		CapacityIDs:           capacityIDs,
+		Nodes:                 nodes,
+		IPRange:               ipRange,
+		AddWGAccess:           addWGAccess,
+		AccessNodeCapacityID:  d.Get("access_node_capacity_id").(uint64),
+		AccessWGConfig:        d.Get("access_wg_config").(string),
+		ExternalIP:            externalIP,
+		ExternalSK:            externalSK,
+		PublicNodeID:          uint32(d.Get("public_node_id").(int)),
+		NodesIPRange:          nodesIPRange,
+		NodeDeploymentID:      nodeDeploymentID,
+		CapacityDeploymentMap: capacityDeploymentID,
+		NodeCapacity:          nodeCapacity,
+		CapacityNode:          capacityNode,
+		Keys:                  make(map[uint32]wgtypes.Key),
+		WGPort:                make(map[uint32]int),
+		APIClient:             apiClient,
+		ncPool:                pool,
+		deployer:              deployer.NewDeployer(apiClient.identity, apiClient.twin_id, apiClient.grid_client, pool, true, nil, string(deploymentDataStr)),
 	}
 	return deployer, nil
 }
 
+// getUniqueNodes gets uniqe nodes list from capacity contracts that the user provided, assigns each node to a capacity contract and vice versa.
+func getUniqueNodes(contractIDs []uint64, nodeCapacityID map[uint32]uint64, capacityNode map[uint64]uint32, cl *apiClient) ([]uint32, error) {
+	nodes := []uint32{}
+	for _, id := range contractIDs {
+		contract, err := cl.substrateConn.GetContract(id)
+		if err != nil {
+			return nil, fmt.Errorf("couldn't get capacity contract with id %d. %w", id, err)
+		}
+		node := uint32(contract.ContractType.CapacityReservationContract.NodeID)
+
+		if isIn[uint64](contractIDs, nodeCapacityID[node]) {
+			// this node is already set to another capacity contract that exists in contarctIDs, no need to reset it
+			continue
+		}
+		if !isIn[uint32](nodes, node) {
+			nodes = append(nodes, node)
+			nodeCapacityID[node] = id
+			capacityNode[id] = node
+		}
+	}
+	sort.Slice(nodes, func(i, j int) bool { return nodes[i] < nodes[j] })
+	return nodes, nil
+}
+
 // invalidateBrokenAttributes removes outdated attrs and deleted contracts
 func (k *NetworkDeployer) invalidateBrokenAttributes(sub *substrate.Substrate) error {
 
-	for node, contractID := range k.NodeDeploymentID {
-		contract, err := sub.GetContract(contractID)
+	for capacity := range k.CapacityDeploymentMap {
+		contract, err := sub.GetContract(capacity)
 		if (err == nil && !contract.State.IsCreated) || errors.Is(err, substrate.ErrNotFound) {
+			node := k.CapacityNode[capacity]
 			delete(k.NodeDeploymentID, node)
 			delete(k.NodesIPRange, node)
 			delete(k.Keys, node)
 			delete(k.WGPort, node)
+			delete(k.NodeCapacity, node)
+			delete(k.CapacityDeploymentMap, capacity)
```

Comment:
shouldn't we make sure the deployment itself is created or not? or is there anyway to get the deployment state and then if it is not ok we delete it?